Library/Homebrew/formula_installer.rb

```diff
@@ -617,10 +617,14 @@ def expand_dependencies_for_formula(formula, inherited_options)
                           (formula.head? || !dependent.latest_version_installed?)
 
       bottle_runtime_version = @bottle_tab_runtime_dependencies.dig(dep.name, "version")
+      bottle_runtime_revision = @bottle_tab_runtime_dependencies.dig(dep.name, "revision")
+      minimum_version = unless bottle_runtime_revision.nil?
+        PkgVersion.new(bottle_runtime_version, bottle_runtime_revision)
+      end
 
       if dep.prune_from_option?(build) || ((dep.build? || dep.test?) && !keep_build_test)
         Dependency.prune
-      elsif dep.satisfied?(inherited_options[dep.name], minimum_version: bottle_runtime_version)
+      elsif dep.satisfied?(inherited_options[dep.name], minimum_version: minimum_version)
```

Comment:
The minimum version will be nil for all packages until the tab is updated sometime after this change. That means that `dep.satisfied?` will only return true here if the latest version of the dependency is installed, right? That seems reasonable.

Reply:
> The minimum version will be nil for all packages until the tab is updated sometime after this change.

If so: I think that's unnecessarily conservative. People _hate_ our "auto-upgrade everything under the sun" behaviour and I don't fancy regressing on it this that far until everything has been revision bumped/reinstalled.

Reply:
The problem is though is we've had 6 bug reports about it in the last week alone, and the issue has gone on for 2 months. I don't think it's conservative when we are currently unable to accurately determine what needs to be upgraded when things are revision bumped - which is very often.

The bug reports won't stop if we just continue on without revision information.

Reply:
One tweak we could do safely is that if the revision of the current version of the dependency is 0 then we can safely continue without revision information from the tab.

I could also backfill some revision information to existing manifests which would solve all concerns here.

Reply:
> One tweak we could do safely is that if the revision of the current version of the dependency is 0 then we can safely continue without revision information from the tab.

This seems good enough to ship this PR 👍🏻 



> I could alternatively backfill some revision information to existing manifests which would solve all concerns here and sounds easy enough to script.

This seems more ideal still but doesn't need to block this 👍🏻 



> So it seems the trade off here is between leaving people with broken installs and back to the original quantity of upgrades for a little while.

I agree. It's a hard trade-off. It's worth noting, though, that the "broken install" behaviour used to be Homebrew behaviour more broadly and: some people preferred it to our (overly) conservative approach of upgrading almost everything always.

It's also not "a little while": it's until these bottle manifests are updated the bottles are going to end up being overly conservative.

The ideal long-term fix for this is gonna be an `abi 1` DSL or similar that we manually incremental and `brew test-bot`/`brew audit` enforces but, until then, I think we should take a middle ground.

Reply:
> One tweak we could do safely is that if the revision of the current version of the dependency is 0 then we can safely continue without revision information from the tab.
> 
> I could alternatively backfill some revision information to existing manifests which would solve all concerns here and sounds easy enough to script.

These options sound reasonable to me.

The main thing we're worried about here is the case when the version is the same between the locally installed package and the package dependency but the revision is higher with the package dependency. Does that mean things should install fine if the locally installed package is referencing a newer version of the same package?

It feels like it should work like this if we don't have the revision. I believe that with the changes in this PR the first and second scenarios in this list would require the package to be upgraded/reinstalled unnecessarily.

| Installed Package | Package Dependency | Latest | Upgrade? |
|---|---|---|---|
| Version 1.0 | Version 1.0, Revision 0 | Version 5.0, Revision 5 | No |
| Version 1.5 | Version 1.0, Revision 3 | Version 5.0, Revision 5 | No |
| Version 2.0 | Version 2.0, Revision 6 | Version 5.0, Revision 5 | Yes |
| Version 2.5 | Version 3.0, Revision 0 | Version 5.0, Revision 5 | Yes |
| Version 5.0 | Version 5.0, Revision 5 | Version 5.0, Revision 5 | Yes |

We could also just load the old package file and get the revision directly each time if it's not in the tab but that sounds worse than just backfilling that information one time.

Reply:
> The ideal long-term fix for this is gonna be an abi 1 DSL 

Yes I agree - I recall mentioning the same idea before somewhere.

---

I will look at writing a backfilling script. I think it's worth it to get this right.

Reply:
Ok the idea I had was to use the `org.opencontainers.image.revision` available in each manifest to get the state of the repository at the time the bottle was built to regenerate the versions.

Unfortunately the `org.opencontainers.image.revision` in most manifests are useless as they use a temporary GitHub Actions merge commit which have long been GCed and lost from the repository.

Reply:
> Does that mean things should install fine if the locally installed package is referencing a newer version of the same package?

I think so, yes, so it would be good to handle that case here and then this PR should be good to merge.

> Ok the idea I had was to use the `org.opencontainers.image.revision` (tap commit hash) available in each manifest to get the state of the repository at the time the bottle was built to regenerate the versions.

Could probably use a timestamp instead?

Alternatively, we just rebottle the formulae with the most dependents/dependencies.